clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -2884,6 +2891,7 @@ private int internalProcessConsumerRecord(
               fatalException.getMessage());
         }
       }
+      maybeReportBatchEndOfIncPushStatus(partitionConsumptionState, serverConfig.getBatchReportEOIPIntervalSecond());
```

Comment:
In the batch function, we don't need to check `completed` status, but only check whether the pending inc push version list is empty or not to reduce the chance the race condition.

Reply:
I think about this actually  - I was also concerning about the race condition last week so I eventually turn to do this in drainer logic - that's why I checked different conditions in the new method
1. If list is empty - nothing to report, quit
2. If list is not empty - 
a) if before COMPLETED, check report interval, if meet then just dump this batch and update the last report TS.
b) after COMPLETED - this is what you mentioned the edge case - it will report the one last batch - and since there will be no new EOIP added to the list, this will be completed.

Do you think it makes sense? 
Like, ideally my vision for all status report, including RTS should happen in the SIT thread, but it will be a large refactor effort and I think we can either do this, or maybe the one as you pointed out, use a map to guard.

Reply:
IIUC, the current strategy doesn't work for all the cases as the batch status update is triggered by a new EOIP, but what if there is no new EOIP, the pending status will be stuck forever, right?

Reply:
Hmm, the method is actually invoked for every record, not for EOIP, so I think it should not hit the corner case from my understanding?

What I think about doing so in SIT periodically actually has a gotcha:
Let's say we accumulate 100 inc pushes before it reports COMPLETED and there is one more active inc push running. there could be the case that that active inc push report EOIP first and then we report this batch of inc push. even with the filtering, there is still chances that filtered versions still overwrite the active EOIP in the replica status history, and active inc push can stuck forever?

Reply:
Yes, the race condition you mentioned could happen if we can't enforce the ordering.

I saw this method: `maybeReportBatchEndOfIncPushStatus`  is only invoked in `processEndOfIncrementalPush`, and if it is invoked upon every message, maybe it should be fine as there will be always heartbeat messages, which can trigger this logic.

Reply:
Uh I think there is a confusion - I think the change in processEndOfIncrementalPush is depending on the config and condition wether it will report directly or add it to the pending report list.
The maybeReportxxxx method is invoked in internalProcessConsumerMethod

Reply:
I see. 
The diff misled me.
Then we should be good.